### PR TITLE
Faster buildInCondition from array of integers

### PR DESCRIPTION
**Pull request by bethrezen**
`buildInCondition` checks each value from an array for it's type(`NULL` or `Expression`). Otherwise it converts value to PDO binding param, which is not effective for numbers.
My fix just skips variables of integer or double type so they are not binded, leaved as-is and are just imploded on line 1108. That's much faster.

### CODE REVIEW

**Comment by bethrezen** [comment]
Tests are failing because of bindings expected for integers in tests

File: framework/db/QueryBuilder.php [hunk 1]
```diff
@@ -1094,7 +1094,7 @@ public function buildInCondition($operator, $operands, &$params)
                 foreach ($value->params as $n => $v) {
                     $params[$n] = $v;
                 }
-            } else {
+            } elseif (gettype($value) !== 'integer' && gettype($value) !== 'double') {
```
Comment (creocoder) [inline framework/db/QueryBuilder.php]: =>

```php
elseif (!is_int($value) && !is_float($value)) {
```

`gettype()` make sence only when you need result as string to compare against some var or to use inside exceptions.

**Comment by klimov-paul** [comment]
I am afraid this is not a good suggestion.
Try to run following code:
```php
var_dump(is_int('aaa123')); // outputs false
var_dump(is_int('123aaa')); // outputs true!!!
```

**Comment by creocoder** [comment]
@klimov-paul You are right, i forget that its php with its [fractal of bad design](http://eev.ee/blog/2012/04/09/php-a-fractal-of-bad-design/) :)

**Comment by creocoder** [comment]
@bethrezen Please change back to original code, sorry for confusion.

**Comment by creocoder** [comment]
@klimov-paul Please check examples by self before suggesting

```php
var_dump(is_int('aaa123')); // outputs false
var_dump(is_int('123aaa')); // outputs false!
```

@bethrezen All fine with suggestion and current code.

**Comment by creocoder** [comment]
@klimov-paul `is_int()` will return true only if variable type is real integer. Look on examples inside php manual and also you can test that:

```php
is_int(23) = bool(true)
is_int('23') = bool(false)
is_int(23.5) = bool(false)
is_int('23.5') = bool(false)
is_int(NULL) = bool(false)
is_int(true) = bool(false)
is_int(false) = bool(false)
```

**Comment by klimov-paul** [comment]
My mistake, I admit.
The integer typecast is unreliable:
```php
var_dump((integer)'123'); // output: int 123
var_dump((integer)'abracadabra123'); // output: int 0
var_dump((integer)'123abracadabra'); // output: int 123 !!!
```
This led me to confusion.

**Comment by creocoder** [comment]
Yes, type casting lead to mistake conclusion :) Ok, anyway i think tests also should be fixed.

**Comment by klimov-paul** [comment]
Well, now I am unsure about float.
Decimal separator at PHP and SQL server may be different.
We have an issue with Sphinx #3513 related to this matter.
As far as I remember Oracle may be sensative to this.

**Comment by creocoder** [comment]
@klimov-paul You are right. I checked original code logic. PR make integers and float out of bind workflow which is probably wrong and may cause side effects. So mine opinion all fine with original code.

**Comment by bethrezen** [comment]
@creocoder maybe we'll just leave `gettype($value) !== integer` ? What can be wrong with integers?

**Comment by klimov-paul** [comment]
High value integers like 15E23 may also have different representation at SQL side. Direct pasting thier PHP representation may lead to SQL parse error.

**Comment by creocoder** [comment]
@bethrezen Last @klimov-paul doubts not about `gettype($value) !== integer` VS `is_int($value)`. We finally find that its equivalent. The trouble is generally is avoid binding itself. Now its clean that we should bind all values regardless of type because not going that way may cause side effects in some cases.

**Comment by bethrezen** [comment]
OK, but I think we should add information on this somewhere(to docs?)
The real-life use-case:
`->where(['in', 'id', $big_array_of_integers]);`
is 10 times slower then
`->where('id in ' . implode(', ', $big_array_of_integers));`

Implode is 10x faster!

BTW, can't find any real-life example, where we will store 15e23 in a php variable.My 64-bit PHP on mac has PHP_INT_MAX = 9223372036854775807 which is lower then 15e23.
Also: 
``` php
php > $i=15e23;
php > var_dump(is_int($i));
bool(false)
php > var_dump(is_int(PHP_INT_MAX));
bool(true)
php > echo gettype($i);
double
php > echo PHP_INT_MAX;
9223372036854775807
```
15e23 IS NOT AN INTEGER!

Do you wanna make yii2 faster or make it compatible with improved SQL syntax?

**Comment by klimov-paul** [comment]
Actually, we have quite an opposite opinion here:
https://github.com/yiisoft/yii/issues/2031

Using parameters binding instead of direct SQL composition may be actually better for the performance at some cases.

**Comment by bethrezen** [comment]
@klimov-paul I'm saying about integers not literals!

**Comment by samdark** [comment]
Performance is better if query is repeated multiple times (except params) for the same connection.

**Comment by samdark** [comment]
In case of persistent connections it could be multiple requests.

**Comment by nineinchnick** [comment]
But only if the number of parameters match.

There are two use cases: fetching records via primary key and others, like fetching multiple records, where the number of keys might change, for example batch operations.

In general, this seems like optimization for a special condition.

**Comment by samdark** [comment]
@nineinchnick of course. Yes, that's kind of optimization. We should be very careful with it.

**Comment by samdark** [comment]
@bethrezen do you have any real benchmarks?

**Comment by bethrezen** [comment]
I see significant performance problem in PDO binded params.
Here we have deal with integers, that are not really needed to be used with binding.
I have 20k id's (integers) and put them to "in" condition.
The query on mysql takes about 0.2s while the pdo binding takes about 35-40 seconds!
If i implode integers into single string as in my PR - there's no PDO binding and no useless 35 seconds!
PDO is slow and we all know it. Therefore we should optimize working with it to ovoid production of useless work for it.

Just say me, WHY we must use PDO binding with integers in `buildInCondition`? I am not talking about other use-cases, model saving or anything else. Just about queries with "in".

**Comment by samdark** [comment]
Are you sure 20K is a real scenario?

**Comment by bethrezen** [comment]
@samdark exchange with 1C, mass updating of prices(+10% for example)
OK, assume that we have smaller count of integers. Anyway, why and for what must we spent all this useless CPU cycles? For ensuring that integer-typed variable stores true integer number? Or what?

**Comment by samdark** [comment]
@bethrezen I agree that it could safely be done but we should benchmark common scenarios i.e. 10—100 IDs.

**Comment by creocoder** [comment]
> The query on mysql takes about 0.2s while the pdo binding takes about 35-40 seconds!

Looks like specific php troubles.

**Comment by bethrezen** [comment]
@creocoder That is PHP-PDO troubles!

**Comment by creocoder** [comment]
@bethrezen Yes, correct, can you provide which version of php was used?

**Comment by bethrezen** [comment]
5.5.21 macosx 
PHP 5.5.21-1+deb.sury.org~trusty+2 - ubuntu 14.04.1 64-bit

**Comment by samdark** [comment]
@bethrezen so what about performance test for 10—100 ids?

@qiangxue what's your opinion about it?

Tests should be fixed if it's considered to be merged.

**Comment by cebe** [comment]
Closing as there is no progress on this and it is unclear how to proceed. If you want to work on this further, please let us know and we'll reopen it.

**Pull request closed without merging**